engine/googletest_helpers.cpp

```diff
@@ -0,0 +1,344 @@
+"[  PASSED  ] 1 test.\n"
+);
+
+/// A test scenario that passes but then exits with non-zero.
+static void
+test_pass_but_exit_failure(void)
+{
+    std::cout << pass_message3;
+    std::exit(70);
+}
+
+
+/// An example incomplete output for `test_timeout`.
+const char incomplete_test[] = (
+"Note: Google Test filter = Suite.incomplete\n"
+"[==========] Running 1 test from 1 test suite.\n"
+"[----------] Global test environment set-up.\n"
+"[----------] 1 test from Suite\n"
+"[ RUN      ] Suite.incomplete\n"
+);
+
+/// A test scenario that times out.
+///
+/// Note that the timeout is defined in the Kyuafile, as the TAP interface has
+/// no means for test programs to specify this by themselves.
+static void
+test_timeout(void)
+{
+    std::cout << incomplete_test;
+
+    ::sleep(10);
+    const fs::path control_dir = fs::path(utils::getenv("CONTROL_DIR").get());
+    std::ofstream file((control_dir / "cookie").c_str());
+    if (!file)
+        fail("Failed to create the control cookie");
+    file.close();
+}
+
+
+}  // anonymous namespace
+
+
+/// Prints out program usage and exits with a non-zero exit code.
+static void
+usage(const char* argv0) {
+    std::cout << "usage: " << argv0 << " "
+              << "[--gtest_color=*] "
+              << "[--gtest_filter=POSITIVE_PATTERNS] "
+              << "[--gtest_list_tests]"
+              << "\n\n"
+              << "This program mocks a googletest test program."
+              << "\n";
+    std::exit(EXIT_FAILURE);
+}
+
+/// Entry point to the test program.
+///
+/// The caller can select which test scenario to run by modifying the program's
+/// basename on disk (either by a copy or by a hard link).
+///
+/// \todo It may be worth to split this binary into separate, smaller binaries,
+/// one for every "test scenario".  We use this program as a dispatcher for
+/// different "main"s, the only reason being to keep the amount of helper test
+/// programs to a minimum.  However, putting this each function in its own
+/// binary could simplify many other things.
+///
+/// \param argc The number of CLI arguments.
+/// \param argv The CLI arguments themselves.  These are not used because
+///     Kyua will not pass any arguments to the plain test program.
+int
+main(int argc, char** argv)
+{
+    using scenario_fn_t = void (*)(void);
+
+    std::map<std::string, scenario_fn_t> scenarios;
+    std::string testcase;
+
+    const char *gtest_color_flag = "--gtest_color=";
+    const char *gtest_filter_flag = "--gtest_filter=";
+    char *argv0 = argv[0];
+
+    bool list_tests = false;
+
+    scenarios["check_configuration_variables"] =
+        test_check_configuration_variables;
+    scenarios["crash"] = test_crash;
+    scenarios["fail"] = test_fail;
+    scenarios["pass"] = test_pass;
+    scenarios["pass_but_exit_failure"] = test_pass_but_exit_failure;
+    scenarios["timeout"] = test_timeout;
+
+    /// NB: this avoids getopt_long*, because its behavior isn't portable.
```

Comment:
What exactly isn't portable? Don't leave readers guessing why you skipped this, because they could try using this interface, seeing it "mostly" work, and submitting a change.

But... Kyua *already* uses `getopt_long` for all of its options parsing and that seems to be fine? See `utils/cmdline/parser.cpp`.

Reply:
*blinks* Huh... I totally overlooked that by accident :D...